### PR TITLE
creeate _preprare_fsdp to pre- prepare fsdp model training

### LINKED ISSUE #3209: The optimizer is not receiving the FSDP model parameters.
### System Info

```Shell
- `Accelerate` version: 1.0.1
- Platform: Linux-5.15.153.1-microsoft-standard-WSL2-x86_64-with-glibc2.31
- `accelerate` bash location: /home/a/anaconda3/envs/trans/bin/accelerate
- Python version: 3.12.7
- Numpy version: 2.1.2
- PyTorch version (GPU?): 2.5.0+cu124 (True)
- PyTorch XPU available: False
- PyTorch NPU available: False
- PyTorch MLU available: False
- PyTorch MUSA available: False
- System RAM: 15.46 GB
- GPU type: NVIDIA GeForce RTX 3070 Laptop GPU
- `Accelerate` default config:
	Not found
```

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [x] My own task or dataset (give details below)

### Reproduction

See code in the pdf
[fsdp_acc.pdf](https://github.com/user-attachments/files/17599827/fsdp_acc.pdf)

### Expected behavior

The optimizer has the FSDP model parameters.

### PR DESCRIPTION
**Pull request by eljandoubi**
Fixes #3209

## Who can review?

- Fully-Sharded Data Parallism: @muellerzr

### CODE REVIEW

**Comment by eljandoubi** [comment]
@muellerzr have you any feedback?

File: src/accelerate/accelerator.py [hunk 1]
```diff
@@ -2075,6 +2079,173 @@ def _prepare_msamp(self, *args, device_placement):
...
+            return args
+
+        # Flattening weights implies that the optimizers have already been processed.
+        if next(next(iter(models.values())).named_parameters())[0].endswith("_flat_param"):
+            return args
+
+        if len(models) != len(optimizers):
+            raise ValueError(
+                f"The number of models ({len(models)}) must be equal to the number of optimizers ({len(optimizers)})."
+            )
```
Comment (BenjaminBossan) [inline src/accelerate/accelerator.py:2105]: Would it make sense to move these checks to the very start of the method?

File: src/accelerate/accelerator.py [hunk 2]
```diff
@@ -2075,6 +2079,173 @@ def _prepare_msamp(self, *args, device_placement):
...
+                            model_layer_group_map[model_idx][name] = group_idx
+                            break  # Exit after the first match to avoid redundant mappings
+
+                    if name in model_layer_group_map[model_idx]:
+                        break
+
+        # Clear parameter lists.
+        for opt in optimizers.values():
+            for group in opt.param_groups:
+                group["params"] = []
```
Comment (BenjaminBossan) [inline src/accelerate/accelerator.py:2164]: Is it better to call `group["params"].clear()`? That would affect references.

**Review by BenjaminBossan — COMMENTED** [review]
Thanks for working on this important issue.

I wonder on the overall approach to solving this issue. Right now, IIUC, the idea is to re-initialize the optimizer(s) using the FSDP-wrapped models. This could potentially be error prone and ideally should be well tested before we merge.

Would it be possible instead to check if this has happened and raise an error, then direct users towards not initializing the optimizer prematurely, and handle the initialization with accelerate so that it's in the right order? Maybe that can't work, but  I think it's worth considering different solutions.

Also, in case this is really the same issue as [this one reported in PEFT](https://github.com/huggingface/peft/issues/2205), it means that it used to work correctly in previous versions of accelerate/transformers. In that case, I wonder what changed that resulted in the issue.

File: src/accelerate/accelerator.py (hunk 1 above)
Comment (eljandoubi) [reply src/accelerate/accelerator.py:2107]: @BenjaminBossan The method! Do you mean `.prepare`? What are the benefits of doing so?

**Comment by eljandoubi** [comment]
@BenjaminBossan  
- You are right; the solution needs more thorough testing. I have conducted a small test in a multi-GPU environment, but it should be tested against other configurations and models.  
- Raising an error when the optimizer is initialized prematurely could be a solution, but two cases emerge:  
  1. If Accelerate initializes the optimizer internally, either it creates a default optimizer with limited flexibility, or the user provides optimizer specifications. This would require external special treatment for FSDP, causing Accelerate to lose its unified interface for distributed training.  
  2. If the user initializes the optimizer post-FSDP setup, Accelerate also loses generalization in this case.  
- For PEFT training, which involves mixing frozen and non-frozen parameters, `use_orig_params=True` must be used. This ensures non-flattened parameters, meaning no changes are needed in the optimizer (I assume).  
- The Transformers Trainer internally uses `delay_optimizer_creation` and creates the optimizer after FSDP wrapping.

File: src/accelerate/accelerator.py (hunk 1 above)
Comment (BenjaminBossan) [reply src/accelerate/accelerator.py:2107]: I meant `_prepare_fsdp`. It is a common pattern to perform all checks as early as possible, so following this makes the code easier to understand for readers. This is especially so if there are early returns.

The reason why it's common is so that we don't do any unnecessary work if the checks fail anyway. In this case, there is no need to determine `models` or `optimizers` if we're going to raise an error later. By skipping the unnecessary work, we ensure faster execution and prevent possibly unwanted side-effects (this might not be relevant here _right now_ but code will change in the future and then it could be true).

**Review by BenjaminBossan — COMMENTED** [review]
Thanks for the updates and explaining your testing and reasoning further.

> Raising an error when the optimizer is initialized prematurely could be a solution, but two cases emerge

These are valid concerns. At the end of the day, it's a trade off and we need to decide which cost we'd rather pay.

> The Transformers Trainer internally uses delay_optimizer_creation and creates the optimizer after FSDP wrapping.

I wonder if this logic could be moved/copied to accelerate.

> For PEFT training, which involves mixing frozen and non-frozen parameters, use_orig_params=True must be used

I don't think this is strictly necessary, we have examples in PEFT with `use_orig_params=False`. But I have to admit I don't know what exactly changes under the hood in FSDP when this parameter is set. Note also that in the linked PEFT issue, I tried setting `use_orig_params=True` and it didn't help.

**Comment by eljandoubi** [comment]
- The [[PyTorch FSDP documentation](https://pytorch.org/docs/stable/fsdp.html)](https://pytorch.org/docs/stable/fsdp.html) mentions that mixed frozen and non-frozen parameters are only supported when `use_orig_params=True`. I believe this is due to the flattening process — we cannot flatten parameters with gradients together with those without. My contribution does not handle the case where `use_orig_params=False` because, in that scenario, the parameters remain unchanged.  
- Another solution would be to adapt an approach similar to the DeepSpeed integration. This involves using a single model and optimizer, and after wrapping the model with `model = FSDP(model)`, we create the optimizer as `optimizer = optimizer_class(model.parameters(), **defaults)`. See #3255

**Comment by BenjaminBossan** [comment]
> * The [PyTorch FSDP documentation](https://pytorch.org/docs/stable/fsdp.html) mentions that mixed frozen and non-frozen parameters are only supported when `use_orig_params=True`. I believe this is due to the flattening process — we cannot flatten parameters with gradients together with those without.

This is true but using the right FSDP auto wrap policy, trainable and frozen parameters should be prevented from being mixed.

> Another solution would be to adapt an approach similar to the DeepSpeed integration.

If this is an option here, it sounds like the more robust solution to me. Let's wait for @muellerzr return to office and get his opinion on this.

**Comment by muellerzr** [comment]
This feels a bit overengineered and the commit to transformers broke a lot of users code with saving checkpoints. For now let's not move forward with this IMO, since otherwise this would imply that the prior change broke all FSDP code, which isn't the case.

**Comment by BenjaminBossan** [comment]
@eljandoubi Could you please check if https://github.com/huggingface/transformers/pull/35212 solved the issue (installing transformers 4.47.1 should contain the fix)?

**Comment by eljandoubi** [comment]
@BenjaminBossan Since the `trainer` delays the creation of the optimizer, this resolves the issue for `transformers`. On the other hand, the core concept of `accelerate` is to pass the model, optimizer, dataloader, and other components simultaneously.

**Comment by BenjaminBossan** [comment]
@eljandoubi Right. As [another user noted](https://github.com/huggingface/accelerate/issues/3256#issuecomment-2558266582), even with `Trainer`, it can still be problematic if users pass their own optimizer instance. I'm not sure if we can find a simpler solution, I think it's worth evaluating the possibilities. @muellerzr do you have a better idea for this situation?
